### DISCUSSION TITLE
Identification flow

### DISCUSSION THREAD

**Discussion by xesjkeee** [directus/directus#5205] [Feature Requests]
Hey 👋 

@rijkvanzanten & @workatease Thank you so much for this [PR](https://github.com/directus/directus/pull/5045)! Thats a huge improvement thing for me ❤️ 

For now directus checks  for `email` field to identify users. So is any way to override field to identify users? I just have cases when users have more than one email, but still have one unique parameter – `username` (login).

How can I change identification flow to check username instead of email and what you thoughts about it?

**Comment by rijkvanzanten** [comment]
Heyaa!


There's currently no way to override that to be another field. That being said, it has come up before (https://github.com/directus/directus/discussions/4237, https://github.com/directus/directus/discussions/4958, and some others) and I do realize that the email address isn't always the primary wanted authentication identifier. 

That being said though.. There's a couple questions: 

What about password reset?
What about verifying if the user is actually who they say they are?
What about SSO?

I'm imagining that we could introduce a "auth identifier field" environment variable that lets you set any arbitrary field in directus_users as the "email" field, which combined with custom fields could mean you could login using the combination <any field> + password. 

(@benhaynes thoughts)

**Reply by xesjkeee** [reply]
Good questions, Rijk 🤔 

In my project I use authentication only via SSO provider, thats mean I haven't users passwords in directus and it closes my worries about password reset and verifying users who they are

I think for my case it would be like the «(o)auth identifier field», but I understand what you mean ... I have no idea how to make it multipurpose not only for oauth

**Reply by rijkvanzanten** [reply]
> I think for my case it would be like the «(o)auth identifier field»

This is another interesting case. Email is something that seems to be generally universally available between most oAuth providers, though I'm not sure how that'd work if you have 3 different oAuth providers that all have different identifiers 🤔

**Reply by rijkvanzanten** [reply]
Also quick PSA: these questions are not meant to shoot down the feature in any way, just want to make sure we can achieve the flexibility we're looking for without making it wildly complicated to setup / break other things in the process 😄

**Reply by benhaynes** [reply]
We should still be able to send a password reset to the email on file if someone identified themselves with a username (or another unique field). Similarly, we could still verify the email on the account separately from the auth credentials.

I think this is an interesting thing to look into more, the idea that you can choose the Directus Users unique field that gets presented as the identified part of auth. Not sure if/how all this would work with SSO, but I understand why this request arises... some interesting use-bases below:

* Username
* Employee ID
* Country SSN
* Disable Normal Auth (eg: SSO Only)

**Comment by xesjkeee** [comment]
@rijkvanzanten @benhaynes 
Hey guys! I appreciate your answers and I want to back to my question. 
Lately, I didn't follow enough deeply for new issues/questions/discussions about similar requests, but I still need it to get the identification of my users who have a few emails.

So, I went to the `/oauth/:provider/callback` router and found the `getEmailFromProfile` function that gets an email from the `profile` object which is a response from an OAuth provider:
https://github.com/directus/directus/blob/02cf1741bbd7cf8c0611465d3b27693872775cbd/api/src/controllers/auth.ts#L313-L323

Let's look inside the function and we'll see the comments:
https://github.com/directus/directus/blob/02cf1741bbd7cf8c0611465d3b27693872775cbd/api/src/utils/get-email-from-profile.ts#L10-L19

Hmm, very interesting [undocumented](https://docs.directus.io/reference/environment-variables/#oauth) environment variable `OAUTH_<PROVIDER>_PROFILE_EMAIL`, isn't it? 🤔😁

Okay, let's continue our adventure. What will you think changes, if I define that variable in the `.env` file? You're right, nothing! Because our `AuthenticationService` service won't check for the defined variable in a database, look at the `whereRaw` method, we have `'email'` constant:
https://github.com/directus/directus/blob/02cf1741bbd7cf8c0611465d3b27693872775cbd/api/src/services/authentication.ts#L55-L72

In finally, I'm not sure how the `profileMap[provider]` works in the
https://github.com/directus/directus/blob/02cf1741bbd7cf8c0611465d3b27693872775cbd/api/src/utils/get-email-from-profile.ts#L19

In my mind, it'll be always `undefined` and due to this, we will get constant `email` (ofc, if we don't define `OAUTH_<PROVIDER>_PROFILE_EMAIL` variable), please correct me if I'm wrong.

I think, if we update
```js
.whereRaw('LOWER(??) = ?', ['email', email.toLowerCase()])
```
to
```js
.whereRaw('LOWER(??) = ?', [env[`OAUTH_${provider.toUpperCase()}_PROFILE_EMAIL`] || 'email', email.toLowerCase()])
```
it'll resolve the general problems about identification. What do you think?

**Reply by rijkvanzanten** [reply]
> Hmm, very interesting undocumented environment variable OAUTH_<PROVIDER>_PROFILE_EMAIL, isn't it? 🤔😁

Whoops! Lemme add that to the reference 👍🏻 

> Okay, let's continue our adventure. What will you think changes, if I define that variable in the .env file? You're right, nothing! Because our AuthenticationService service won't check for the defined variable in a database, look at the whereRaw method, we have 'email' constant:

It will change _something_ though, it'll look for that new path in the object received by the oAuth provider. So if the oAuth provider responds with

```json
{
  "data": {
    "user_profile": {
      "email": "rijk@directus.io"
    }
  }
}
```

you can control where Directus will look for the email address in that object structure by setting `OAUTH_FACEBOOK_PROFILE_EMAIL` to something like `data.user_profile.email` (using `facebook` as my example provider here).

This does _not_ control what field in the users collection holds the email address, as that column is called `email`.

> In finally, I'm not sure how the profileMap[provider] works in the

The code provided:

```js
profileMap[provider] || env[`OAUTH_${provider.toUpperCase()}_PROFILE_EMAIL`] || 'email'; 
```

basically is a fallback chain. If we don't have a known path for the profile email, we'll fallback to the user provided "PROFILE_EMAIL" setting. If that isn't provided, we'll fall back to a generic "email" field, which works with a lot of providers.

> I think if we update [...] to [...] it'll resolve the general problems about identification. What do you think?

I think you're mistaken here. By changing that `email` in the query to the value of the environment variable, we'd end up mistaking the column in the directus_users table that holds the stored email with the path in the profile JSON object that comes from the oauth provider. We wouldn't want to query `SELECT data.user_profile.email FROM directus_users` for example.

**Reply by rijkvanzanten** [reply]
> Hmm, very interesting undocumented environment variable OAUTH__PROFILE_EMAIL, isn't it? 🤔😁
> > Whoops! Lemme add that to the reference 👍🏻

Hey! It's in the docs already 🙂 

![CleanShot 2021-08-18 at 15 27 40@2x](https://user-images.githubusercontent.com/9141017/129960028-91de1738-b539-475d-97a0-09a8b8712687.png)

https://docs.directus.io/reference/environment-variables/#oauth

**Reply by xesjkeee** [reply]
> Hey! It's in the docs already 🙂

But... `OAUTH_<PROVIDER>_PROFILE_URL` is not `OAUTH_<PROVIDER>_PROFILE_EMAIL` 🙂

I think, somebody mistook `PROFILE_URL` with `PROFILE_EMAIL`. And I didn't find any references in the code for the documented `OAUTH_<PROVIDER>_PROFILE_URL`:

<img width="357" alt="Screen" src="https://user-images.githubusercontent.com/17751886/129968503-9154518c-ebd8-479e-920b-b26b43525454.png">

> I think you're mistaken here. By changing that email in the query to the value of the environment variable, we'd end up mistaking the column in the directus_users table that holds the stored email with the path in the profile JSON object that comes from the oauth provider. We wouldn't want to query SELECT data.user_profile.email FROM directus_users for example.

Yeah, you're right, now I understand that my way was wrong.

So, what do you think about adding a new variable like an `OAUTH_<PROVIDER>_IDENTIFICATION_KEY` which helps control there:
https://github.com/directus/directus/blob/67e9b94f106456f58eafe0e14c409ff614d35cf6/api/src/services/authentication.ts#L68-L72

to handle the identification of incoming profiles from OAuth providers with the existing user in a database?

I'm ready to create PR for that, but I don't know that you want it, maybe you have another vision for that.

**Reply by rijkvanzanten** [reply]
> But... OAUTH_<PROVIDER>_PROFILE_URL is not OAUTH_<PROVIDER>_PROFILE_EMAIL 🙂

Ahhh good point.

> So, what do you think about adding a new variable like an OAUTH_<PROVIDER>_IDENTIFICATION_KEY which helps control there:

Totally, though it shouldn't be locked to oAuth. @aidenfoxx just implemented this same exact thing in https://github.com/directus/directus/pull/6942 as preparation for integrating LDAP 🙂

**Reply by xesjkeee** [reply]
Good news! But there is a check only for `email` 😕

<img width="819" alt="Screen" src="https://user-images.githubusercontent.com/17751886/129985881-978b040b-7124-44e4-915b-971a42c43e75.png">

**Reply by CallMeLaNN** [reply]
@xesjkeee I have the same concern too but currently I workaround by giving `<userId>@mydomain.com` to satisfy the constrain and uniqueness. You can make it as `<username>@mydomain.com` so it will be unique. Since you use external authentication, I believe you doesn't need email sending feature.

BTW, `OAUTH_<PROVIDER>_PROFILE_URL` is coming from [`grant` `profile_url`](https://github.com/simov/grant/#configuration-values). Directus makes `OAUTH_<PROVIDER>_*` dynamic config to be passed IMO.

I also just start [a discussion](https://github.com/directus/directus/discussions/7442) (see today comment on identification) on a bigger picture but try to make it smooth to implement it. I'm not sure if you meant to have multiple providers with different emails merging to single account like StackExchange.

---

 @rijkvanzanten 

> @aidenfoxx just implemented this same exact thing in #6942 as preparation for integrating LDAP 🙂

Aah, good point, LDAP was not in my search keyword, I'll go trough this.

**Reply by CallMeLaNN** [reply]
@xesjkeee @rijkvanzanten 

Another option without `OAUTH_<PROVIDER>_IDENTIFICATION_KEY` is to let `user.id` as string and use that just like any other collection id for manual string. I already scan trough all the source code. IMO no harm to change that except db like Postgress UUID format constrain because my user id coming from mongo ObjectId.

However this way doesn't scale much, we can't have multiple providers like we already have. That is where mapping 1:many between `user.id` and external id is good. Additionally we can't rely on username, some provider can change email while other can change username.

**Reply by aidenfoxx** [reply]
@CallMeLaNN @xesjkeee  The changes made to auth currently don't affect the oauth flow too much because it wasn't the focus of the PR. The real intention was to modify the core to allow alternative login methods other than just email. I did discuss with Rijk about how we'd want to move oauth to this new system, but that was planned for after #6942 was merged.

On that note, #6942 does add a new `identifier` field to users that could be used as part of SSO. Instead of checking against an email value, oauth could check against `identifier`, and the `identifier` type could be configured with `OAUTH_<PROVIDER>_IDENTIFICATION_KEY`.

EDIT: Also, there is no LDAP code in that PR. The LDAP stuff is on a different branch.

**Reply by xesjkeee** [reply]
@aidenfoxx @CallMeLaNN thanks, guys!
It unconditionally valuable feature in that PR that allows continuing work on the custom identification key.
I look forward to merging it to the main branch 🤗

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
